projects/schematics/src/migrations/4_0/constructor-deprecations/data/modal.service.migration.ts

```diff
@@ -0,0 +1,25 @@
+import {
+  ANGULAR_CORE,
+  APPLICATION_REF,
+  FEATURE_CONFIG_SERVICE,
+  MODAL_SERVICE,
+  NGB_MODAL,
+  NG_BOOTSTRAP,
+  SPARTACUS_STOREFRONTLIB,
+} from '../../../../shared/constants';
+import { ConstructorDeprecation } from '../../../../shared/utils/file-utils';
+
+export const MODAL_SERVICE_MIGRATION: ConstructorDeprecation = {
+  // projects/storefrontlib/src/shared/components/modal/modal.service.ts
+  class: MODAL_SERVICE,
+  importPath: SPARTACUS_STOREFRONTLIB,
+  deprecatedParams: [
+    { className: NGB_MODAL, importPath: NG_BOOTSTRAP },
+    { className: APPLICATION_REF, importPath: ANGULAR_CORE },
+    { className: FEATURE_CONFIG_SERVICE, importPath: SPARTACUS_STOREFRONTLIB },
+  ],
+  removeParams: [
+    { className: FEATURE_CONFIG_SERVICE, importPath: SPARTACUS_STOREFRONTLIB },
+  ],
+  addParams: [{ className: APPLICATION_REF, importPath: ANGULAR_CORE }],
+};
```

Comment:
Could you add second migration for situation where customer didn't provided ApplicationRef and FeatureConfigService?


Also in this migration you don't need `addParam`, as this migration will work for customers that provided both of those services, so only the remove was needed.